### PR TITLE
Trev8939/subnetwork trace

### PR DESCRIPTION
**Pull request by TADraeseke**
Hey all! Can you have a review of this please? Confusing bit of API so please ask questions about the code and feel free to offer suggestions to simplify any confusing bits.

Thanks!

### CODE REVIEW

File: kotlin/configure-subnetwork-trace/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    <uses-permission android:name="android.permission.INTERNET"/>
+
+    <application
+        android:allowBackup="true"
+        android:icon="@mipmap/ic_launcher"
+        android:label="@string/app_name"
+        android:roundIcon="@mipmap/ic_launcher_round"
+        android:supportsRtl="true"
+        android:theme="@style/AppTheme">
+        <activity android:name=".MainActivity">
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/AndroidManifest.xml]: within `activity`, you could add `android:screenOrientation="landscape">` so that it doesn't need to be called in the MainActivity.kt class

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+  private var sources: List<UtilityNetworkAttribute>? = null
+  private var operators: Array<UtilityAttributeComparisonOperator>? = null
+  private var startingLocation: UtilityElement? = null
+  private var values: List<CodedValue>? = null
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: this can be hidden away in the AndroidManifest.xml: see code suggestion above!

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+                  override fun onItemSelected(
+                    parent: AdapterView<*>?,
+                    view: View?,
+                    position: Int,
+                    id: Long
+                  ) {
+                    (sources[sourceSpinner.selectedItemPosition])
+                    onComparisonSourceChanged(sources[sourceSpinner.selectedItemPosition])
+                  }
+
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: can remove white space here

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+                  override fun onNothingSelected(parent: AdapterView<*>?) {}
+                }
+              }
+            }
+
+        operators = UtilityAttributeComparisonOperator.values().also { operators ->
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: a comment for what is going on here would be good

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        }
+      }
+    }
+  }
+
+  /**
+   * When a comparison source attribute is chosen check if it's a coded value domain and, if it is,
+   * present a spinner of coded value domains. If not, show the correct UI view for the utility
+   * network attribute data type.
+   */
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: missing @param for `attribute`

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+    (attribute.domain as? CodedValueDomain)?.let { codedValueDomain ->
+      // update the list of coded values
+      values = codedValueDomain.codedValues
+      // show the values spinner
+      setVisible(valuesBackgroundView.id)
+      // update the values spinner adapter
+      valuesSpinner.adapter = ArrayAdapter<String>(
+        applicationContext,
+        android.R.layout.simple_spinner_item,
+        // add the the coded values from the coded value domain to the values spinner
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: duplicate "the" here

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+      }
+      UtilityNetworkAttribute.DataType.INTEGER -> {
+        // show the edit text only allowing for integer input
+        valuesEditText.inputType = InputType.TYPE_CLASS_NUMBER
+        setVisible(valuesEditText.id)
+      }
+      else -> {
+        Log.e(TAG, "Unexpected utility network attribute data type.")
+      }
+
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: can remove white space here

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+    }
+  }
+
+  /**
+   * Show the given UI view and hide the others which share the same space.
+   */
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: missing @param here for `id`

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+        valueBooleanButton.visibility = View.VISIBLE
+        valuesBackgroundView.visibility = View.INVISIBLE
+        valuesEditText.visibility = View.INVISIBLE
+      }
+    }
+  }
+
+  /**
+   * Add a new barrier condition to the trace options.
+   */
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: missing @param for `view` here

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+          expression = UtilityTraceOrCondition(otherExpression, expression)
+        }
+        traceConfiguration.traversability.barriers = expression
+        expressionTextView.text = expressionToString(expression)
+      }
+    }
+  }
+
+  /**
+   * Run the network trace with the parameters and display the result in an alert dialog.
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: missing @param for `view` here

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+    requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
+
+    exampleTextView.movementMethod = ScrollingMovementMethod()
+
+    // create a utility network and wait for it to finish to load
+    utilityNetwork.loadAsync()
+    utilityNetwork.addDoneLoadingListener {
+      if (utilityNetwork.loadStatus == LoadStatus.LOADED) {
+        // populate spinners for network attribute comparison
+        sources =
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: as we discussed, some more comments through here to help readability would be great, along with possibly separating the UI from our API into a different method if that seems doable.

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+    // if source tier doesn't contain a trace configuration, create one
+    val traceConfiguration = sourceTier?.traceConfiguration ?: UtilityTraceConfiguration().apply {
+      // if the trace configuration doesn't contain traversability, create one
+      traversability ?: UtilityTraversability()
+    }
+
+    // NOTE: You may also create a UtilityCategoryComparison with UtilityNetworkDefinition.Categories and UtilityCategoryComparisonOperator
+    sources?.get(sourceSpinner.selectedItemPosition)?.let { attribute ->
+      operators?.get(operatorSpinner.selectedItemPosition)?.let { attributeOperator ->
+        // NOTE: You may also create a UtilityCategoryComparison with UtilityNetworkDefinition.Categories and UtilityCategoryComparisonOperator
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: duplicate comments here: though can you help me understand why we are telling the user this at this point in the code? From the comment I was expecting the code below to show a UtilityCategoryComparison being created, though I don't see that happening.

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+    } catch (exception: Exception) {
+      ("Error during trace operation: " + exception.message).also {
+        Toast.makeText(this, it, Toast.LENGTH_LONG).show()
+        Log.e(TAG, it)
+      }
+    }
+  }
+
+  /**
+   * Convert the given UtilityTraceConditionalExpression into a string.
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: missing @ param here for `expression`

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+    initialExpression?.let {
+      val traceConfiguration = sourceTier?.traceConfiguration as UtilityTraceConfiguration
+      traceConfiguration.traversability.barriers = it
+      expressionTextView.text = expressionToString(it)
+    }
+  }
+
+  /**
+   * Convert the given value into the correct Kotlin data type by using the attribute's data type.
+   */
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: missing @ param for `otherValue` and `dataType`

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt (hunk 11 above)
Comment (TADraeseke) [reply kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: I'm hoping the added comments and .also block help indicate that its UI stuff happening with the populating of the sources list

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt (hunk 12 above)
Comment (TADraeseke) [reply kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: Hmm yeah, this is from the common design--I've just scrapped it. It's confusing to me too

**Comment by TADraeseke** [comment]
Thanks @Rachael-E I've tried to implement those fixes, but please take another look when you can :)

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+
+    exampleTextView.movementMethod = ScrollingMovementMethod()
+
+    // create a utility network and wait for it to finish to load
+    utilityNetwork.loadAsync()
+    utilityNetwork.addDoneLoadingListener {
+      if (utilityNetwork.loadStatus == LoadStatus.LOADED) {
+        // create a list of utility network attributes whose system is not defined
+        sources = utilityNetwork.definition.networkAttributes.filter { !it.isSystemDefined }
+          .also { sources ->
```
Comment (hud10837) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: Personally I'm finding it difficult to follow the nesting here...is it necessary to use the also here? If not, it might be a bit more legible without it:

```// create a list of utility network attributes whose system is not defined
sources = utilityNetwork.definition.networkAttributes.filter { !it.isSystemDefined }

sources.sourceSpinner.apply {
  adapter = ArrayAdapter<String>(
    applicationContext,
    android.R.layout.simple_spinner_item,
    sources.map { it.name })

  onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
    override fun onItemSelected(
      parent: AdapterView<*>?,
      view: View?,
      position: Int,
      id: Long
     ) {
      (sources[sourceSpinner.selectedItemPosition])
     onComparisonSourceChanged(sources[sourceSpinner.selectedItemPosition])
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {}
 }
```

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+          // if there's a coded value domain name
+          val codedValueDomainName = codedValueDomain.codedValues.first {
+            convertToDataType(it.code, expression.networkAttribute.dataType) ==
+                convertToDataType(expression.value, expression.networkAttribute.dataType)
+          }.name
+          return expression.networkAttribute.name + " " + expression.comparisonOperator + " " + codedValueDomainName
+        }
+          ?: return expression.networkAttribute.name + " " + expression.comparisonOperator + " " + (expression.otherNetworkAttribute?.name
+            ?: expression.value)
+      }
```
Comment (hud10837) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: ^ This is pretty complex, took me a while to parse out. Not sure if there's a way to separate it out. Or perhaps you could take the first part of the return string out so that the code is a little more compact? ie. 
```
val expressionAttributeAndComparison = expression.networkAttribute.name + " " + expression.comparisonOperator + " "

...

return expressionAttributeAndComparison + codedValueDomainName} ?: expressionAttributeAndComparison + //etc..

```

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+
+  /**
+   * Reset the current barrier condition to the initial expression
+   * "Operational Device Status EQUAL Open".
+   *
+   * @param view of the rest button
+   */
+  fun reset(view: View) {
+    initialExpression?.let {
+      val traceConfiguration = sourceTier?.traceConfiguration as UtilityTraceConfiguration
```
Comment (hud10837) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: I'm not sure about casts yet, but you used `as?` in all the previous casts

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt (hunk 15 above)
Comment (TADraeseke) [reply kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: Hmm, I'm trying to work around it, but the .also scope provides a non-null `sources`

Comment (hud10837) [reply kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: That makes a lot of sense. I'm looking at it now and it's not so hard to follow, but maybe I'm just getting used to these scoping functions now. I haven't used .also too much yet

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 18]
```diff
@@ -293,15 +296,20 @@ class MainActivity : AppCompatActivity() {
       }
       // when the expression is an attribute comparison expression
       is UtilityNetworkAttributeComparison -> {
+        // the name and comparison operator of the expression
+        val networkAttributeNameAndOperator =
+          expression.networkAttribute.name + " " + expression.comparisonOperator + " "
+        // the
```
Comment (hud10837) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: Think you dropped a comment here

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 19]
```diff
@@ -0,0 +1,368 @@
...
+      // when the expression is a category comparison expression
+      is UtilityCategoryComparison -> {
+        return expression.category.name + " " + expression.comparisonOperator
+      }
+      // when the expression is an attribute comparison expression
+      is UtilityNetworkAttributeComparison -> {
+        // the name and comparison operator of the expression
+        val networkAttributeNameAndOperator =
+          expression.networkAttribute.name + " " + expression.comparisonOperator + " "
+        // the
```
Comment (TADraeseke) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]:
```suggestion
        // check whether the network attribute has a coded value domain
```

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt (hunk 18 above)
Comment (TADraeseke) [reply kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: good spot, thanks!

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt (hunk 15 above)
Comment (hud10837) [reply kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: I think after our discussion yesterday afternoon and looking at the activity as a whole, it makes sense. There's a pretty consistent pattern of
```
var myFish = Tuna().[scopingfxn] { fishy ->
  fishy.tank = this
  fishy.taste = Flavor.DELICIOUS
}
```
I guess scoping functions in kotlin let us block together all the kind of initialization code for objects during their creation? I can see the case for both the Java style of "each line does one action" versus this style of "group similar actions into one place". Maybe that's more of a mindset thing to get used to with kotlin?

File: kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt [hunk 20]
```diff
@@ -0,0 +1,368 @@
...
+      ("Error during trace operation: " + e.message).also {
+        Toast.makeText(this, it, Toast.LENGTH_LONG).show()
+        Log.e(TAG, it)
+      }
+    }
+  }
+
+  /**
+   * Convert the given UtilityTraceConditionalExpression into a string.
+   */
```
Comment (Rachael-E) [inline kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt]: looks like the @ param is still missing for `expression` here
